int-set/src/lib.rs

```diff
@@ -0,0 +1,294 @@
+//! A fast, efficient, sparse, & ordered unsigned integer (u32) bit set which is invertible.
+//!
+//! The bitset is implemented using fixed size pages which allows it to compactly
+//! represent sparse membership. However, the set excels when set members are typically
+//! clustered together. For example when representing glyph id or unicode codepoint values
+//! in a font.
+//!
+//! The set can have inclusive (the set of integers which are members) or
+//! exclusive (the set of integers which are not members) membership. The
+//! exclusive/inverted version of the set is useful for patterns such as
+//! "keep all codepoints except for {x, y, z, ...}".
+//!
+//! When constructing a new IntSet from an existing lists of integer values the most efficient
+//! way to create the set is to initialize it from a sorted (ascending) iterator of the values.
+
+mod bitpage;
+mod bitset;
+
+use bitset::BitSet;
+use std::ops::RangeInclusive;
+
+/// A fast & efficient invertible ordered set for small (up to 32-bit) unsigned integer types.
+#[derive(Clone, Debug)]
+pub struct IntSet<T>(Membership<T>);
+
+#[derive(Clone, Debug)]
+enum Membership<T> {
+    /// Records a set of integers which are members of the set.
+    Inclusive(BitSet<T>),
+
+    /// Records the set of integers which are not members of the set.
+    Exclusive(BitSet<T>),
+}
+
+impl<T: Into<u32> + Copy> Default for IntSet<T> {
+    fn default() -> IntSet<T> {
+        IntSet::empty()
+    }
+}
+
+impl<T: Into<u32> + Copy> IntSet<T> {
+    /// Create a new empty set (inclusive).
+    pub fn empty() -> IntSet<T> {
+        IntSet(Membership::Inclusive(BitSet::empty()))
+    }
+
+    /// Create a new set which contains all integers (exclusive).
+    pub fn all() -> IntSet<T> {
+        IntSet(Membership::Exclusive(BitSet::empty()))
+    }
+
+    /// Return the inverted version of this set.
+    pub fn invert(&mut self) {
```

Comment:
these methods (and any others that don't have the `T` param? can probably go in the impl block that has no type constraints?

Reply:
BitSet::empty() requires that T hash Into and Copy

Reply:
Nm, got it working.